### ISSUE TITLE
`no-unnecessary-callback-wrapper` false positive (mismatched args)

### ISSUE THREAD

**Issue by nchen63** [palantir/tslint#2430]
#### TypeScript code being linted

```ts
function a(arg1: string, arg2?: number) {
}

function b(callback: (arg1: string, arg2?: string) => void) {
}

b((arg1) => a(arg1)); // rule warns on this...
b(a);                 // ...but this isn't valid

```

with `tslint.json` configuration:

```json
"no-unnecessary-callback-wrapper": true
```

#### Actual behavior
If the 2nd args are optional and different types, the rule flags this, but it can't be fixed by unwrapping

#### Expected behavior
Shouldn't warn, but I don't think this can be fixed without type checking

**Comment by danvk** [comment]
The classic example of this is `map` and `parseInt`:

```
> ['010', '010', '010'].map(x => parseInt(x))
[10, 10, 10]
> ['010', '010', '010'].map(parseInt)
[10, NaN, 2]
```

**Comment by felixfbecker** [comment]
Same problem

![image](https://user-images.githubusercontent.com/10532611/29340470-fda3b70e-81d3-11e7-9cc5-a291912a47f1.png)

![image](https://user-images.githubusercontent.com/10532611/29340481-0e7695c4-81d4-11e7-80d0-7c1af83d5df0.png)

**Comment by mad-gooze** [comment]
This rule breaks several window methods, for example:
```
const foo = {
    bar: (cb) => setTimeout(cb)
}
``` 
which becomes 
```
const foo = {
    bar: setTimeout
}
``` 
And `foo.bar()` throws `TypeError: illegal invocation`.

This thing happens with all window methods like `setTimeout`, `clearTimeout`, etc.

**Comment by csvn** [comment]
I made a small reproducement repo for a simplified case of something we ran into in our code.

```typescript
await new Promise<void>(res => {
  document.addEventListener('click', () => res());
});
```

Replacing `() => res()` with `res` results in a Typescript error (`Event` and `void` is not compatible for the `Promise`).

https://github.com/csvn/tslint-issue-no-unnecessary-callback-wrapper

**Comment by Springrbua** [comment]
Is there a workaround for this issue?

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
